### PR TITLE
Feat/filterconsensus called reads

### PR DESCRIPTION
**Pull request by ashwini06**
### This PR is a combination of changes:

Added:

- [x] Bash script to filter Consensuscalled reads based on XZ tag. By default, the parameter `min read` is to set to `'3,1,1'` for filtering. 
       - Helps to reduce the number of false-positives from variant-calling. It still needs to identify the best `minreads` settings and check for recall/precision values.

- [x] Refactoring of `sentieon_consensuscall` rule in UMI_workflow
         - Restructured snakemake rule structure and defining the names. Updated cluster.json for the correct rule name
         - Removed the declaration of parameter values in rules. Instead, moved declaration to constant.py,  

## Screenshots
f1
f2

## Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

**Comment by hassanfa** [comment]
Let me know when it is ready to review.

**Comment by ashwini06** [comment]
> Let me know when it is ready to review.

@hassanfa Now it's ready for review

File: BALSAMIC/assets/scripts/FilterDuplexUMIconsensus.sh [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+output_bam=$2
+min_reads=$3
+
+# min_reads has three values, e.g. 2,1,1
+# The first value is applies to the total group size,
+# the second value to one single-strand consensus, 
+# and the last value to the other single-strand consensus.
+
+samtools view -h $input_bam | \
+awk -v MinR=$min_reads -v OFS="\t" '
```
Comment (hassanfa) [inline BALSAMIC/assets/scripts/FilterDuplexUMIconsensus.sh:21]: Could we add this as awk script? We can then move rest of it to snakemake shell. 

Reason is, this script has to be moved into a python, R, or Go script for maintainability in future.

File: BALSAMIC/config/cluster.json [hunk 2]
```diff
@@ -186,5 +186,21 @@
 	"vardict_tumor_only": {
 		"time": "10:00:00",
 		"n": 10
+	},
+	"sentieon_umi_align": {
```
Comment (hassanfa) [inline BALSAMIC/config/cluster.json:190]: Do UMI rules need 24:00:00? Let's reduce this to a lower number. What do you recommend?

File: BALSAMIC/snakemake_rules/umi/generate_AF_tables.rule [hunk 3]
```diff
@@ -1,27 +1,36 @@
...
 # Generate tables for AF scatterplots
 rule calculate_AF:
     input:
-        vcf = vcf_dir + '{sample}.{var_caller}.umi.vcf.gz'
+        vcf = vcf_dir + '{sample}.{var_caller}.consensusfiltered.vcf.gz'
     output:
-        AF = table_dir + '{sample}.{var_caller}.umi.AFtable.txt'
+        AF = table_dir + '{sample}.{var_caller}.consensusfiltered.AFtable.txt'
+    benchmark:
+        benchmark_dir + '{sample}.{var_caller}.AFcalculate.tsv'
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/umi/generate_AF_tables.rule:16]: Could we have rule name first for benchmark file.

File: BALSAMIC/snakemake_rules/umi/sentieon_consensuscall.rule [hunk 4]
```diff
@@ -2,55 +2,98 @@
 # coding: utf-8
 
 
+from BALSAMIC.utils.rule import get_conda_env
+from BALSAMIC.utils.rule import get_script_path
 from BALSAMIC.utils.rule import get_threads
+from BALSAMIC.utils.models import UMIworkflowConfig
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/umi/sentieon_consensuscall.rule:8]: Wonderful!

File: BALSAMIC/utils/models.py [hunk 5]
```diff
@@ -480,3 +480,18 @@ def validate_path(cls, value, values, **kwargs):
                 output_value = value
 
         return output_value
+
+
+class UMIworkflowParams(BaseModel):
```
Comment (hassanfa) [inline BALSAMIC/utils/models.py:485]: Please add attributes to docstring as well.

File: BALSAMIC/utils/models.py [hunk 6]
```diff
@@ -480,3 +480,18 @@ def validate_path(cls, value, values, **kwargs):
...
+    """This class defines the params settings used as constants in UMI workflow """
+    align_format: str
+    filter_minreads: str = '3,1,1'
+    tag: str = 'XR'
+    align_header: str
+    align_intbases: int
+
+class UMIworkflowConfig(BaseModel):
```
Comment (hassanfa) [inline BALSAMIC/utils/models.py:493]: Please extend docstring.

**Review by hassanfa — CHANGES_REQUESTED** [review]
✨✨
Some minor comments.

**Comment by hassanfa** [comment]
Forgot to mention, please also add tests for two new models you've added in https://github.com/Clinical-Genomics/BALSAMIC/blob/master/tests/utils/test_models.py

File: BALSAMIC/assets/scripts/FilterDuplexUMIconsensus.sh (hunk 1 above)
Comment (ashwini06) [reply BALSAMIC/assets/scripts/FilterDuplexUMIconsensus.sh]: yes changed to awk script.

File: BALSAMIC/snakemake_rules/umi/generate_AF_tables.rule (hunk 3 above)
Comment (ashwini06) [reply BALSAMIC/snakemake_rules/umi/generate_AF_tables.rule]: Fixed

**Review by hassanfa — APPROVED** [review]
Very nice! ⭐  ⭐  I liked addition of the parameters to constants. At some point we need to move all rule specific params to constants.

**Comment by ashwini06** [comment]
> Very nice! ⭐ ⭐ I liked addition of the parameters to constants. At some point we need to move all rule specific params to constants.

Thank you :) Sure, will fix in other PR

**Pull request merged**